### PR TITLE
Refactor puppet artifacts to match with second IAM deployment pattern

### LINKED ISSUE #37: Refactor puppet artifacts to match with standard IAM deployment patterns of IS 5.4.0
**Description:**
Puppet artifacts should adhere to following standard IAM deployment patterns.
Pattern-1: HA clustered deployment of WSO2 Identity Server
Pattern-2: HA clustered deployment of WSO2 Identity Server with WSO2 Identity Analytics

**Affected Product Version:**
IS 5.4.0

### PR DESCRIPTION
**Pull request by malithie**
## Purpose
Resolves #37

## Goals
Refactor wso2is_analytics module and add hiera with respect to IAM deployment pattern-2 in wso2is and wso2is_analytics modules.

## Approach
- Refactor templates and configurations to match with the second IAM deployment pattern for wso2is and wso2is_analytics puppet modules
- Update readme files

## Test environment
JDK 8
Ubuntu 14.04
Puppet 3.8.7
Tested in puppet master agent setup

### CODE REVIEW

File: wso2is/hieradata/dev/wso2/wso2is/pattern-2/README.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    e.g:
+     ```
+     wso2::file_list:
+       - "repository/components/lib/%{hiera('wso2::datasources::mysql::connector_jar')}"
+    ```
+3. Configure a clustering scheme
+
+    Configure a clustering scheme under ```wso2::clustering```. By default, WKA related configurations are added. 
+    Update the member list under WKA scheme in [default.yaml](default.yaml) file with the ip addresses of the members
+     participating in the cluster.
```
Comment (DilanUA) [inline wso2is/hieradata/dev/wso2/wso2is/pattern-2/README.md]: check for formatting

File: wso2is/hieradata/dev/wso2/wso2is/pattern-2/README.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+   Configure a JDBC user store, LDAP or an Active Directory, with connection URL, connection username and password.
+   e.g:
+   ```
+   wso2::user_store:
+     user_store_type: jdbc
+     user_store_properties:
+       jdbc:
+         driver_class_name: "%{hiera('wso2::datasources::mysql::driver_class_name')}"
+         url: jdbc:%{hiera('wso2::rdbms::engine')}://%{hiera('wso2::rdbms::hostname')}:%{hiera('wso2::rdbms::port')}/WSO2_USER_DB?autoReconnect=true&amp;useSSL=false
+         username: "%{hiera('wso2::datasources::mysql::username')}"
```
Comment (DilanUA) [inline wso2is/hieradata/dev/wso2/wso2is/pattern-2/README.md]: check for formatting

File: wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+The hiera YAML in this directory can be used to setup WSO2 Identity Analytics Server nodes participating in the 
+above deployment.
+Default configurations in the hiera YAML is to setup a single node deployment of WSO2 Identity Analytics Server.
+
+Follow the given steps below, to configure hiera YAML with respect to the deployment.
+
+1. Update hostname
+
+    Update ```wso2::hostname``` and ```wso2::mgt_hostname``` in the [default.yaml](default.yaml) file with the DNS name 
+    of the load balancer. If the environment does not have a load balancer, then make sure to add the hostname mapping to
```
Comment (DilanUA) [inline wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md]: check for formatting

File: wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    other connection URL query parameters, connection username against  ```username``` and password against 
+    ```password``` property keys. Datasources defined in this hiera file are below.
+    * Datasource in repository/conf/master-datasources.xml file:
+    
+    A single datasource can be configured for registry and user management
+    e.g:
+    ```
+    wso2::master_datasources:
+      analytics_ds:
+          name: WSO2_ANALYTICS_DS
```
Comment (DilanUA) [inline wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md]: How can this be an example for registry and user management ?

File: wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    domain: is.analytics.wso2.domain
+    local_member_host: "%{::ipaddress}"
+    local_member_port: 4000
+    sub_domain: mgt
+    
+    membership_scheme: wka
+    wka:
+     members:
+       -
+         hostname: 192.168.48.194
```
Comment (DilanUA) [inline wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md]: why %{::ipaddress} not used here ?

File: wso2is_analytics/manifests/params.pp [hunk 6]
```diff
@@ -160,14 +161,18 @@
       }
     }
 
+    $analytics_event_store_datasource = 'wso2_analytics_event_store_db'
```
Comment (DilanUA) [inline wso2is_analytics/manifests/params.pp]: why the key refers to a data source and value refers to a database ?

File: wso2is_analytics/manifests/params.pp [hunk 7]
```diff
@@ -160,14 +161,18 @@
...
+
+    $analytics_processed_data_store_datasource = 'wso2_analytics_processed_data_store_db'
```
Comment (DilanUA) [inline wso2is_analytics/manifests/params.pp]: why the key refers to a data source and value refers to a database ?

**Review by DilanUA — CHANGES_REQUESTED** [review]

File: wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md (hunk 4 above)
Comment (malithie) [reply wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md]: 'datasource' element under 'wso2::user_management' and 'wso2::registry_instances' points to the datasource defined here, which are used in The registry.xml and user-mgt.xml templates

File: wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md (hunk 5 above)
Comment (malithie) [reply wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md]: This is the read me entry which specifically ask to configure nodes participating in the cluster

File: wso2is_analytics/manifests/params.pp (hunk 6 above)
Comment (malithie) [reply wso2is_analytics/manifests/params.pp]: That's the default configuration. All product default configurations have 'db's.
If we are to change this point we need to change all IMO

File: wso2is/hieradata/dev/wso2/wso2is/pattern-2/README.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
+# WSO2 Identity Server Pattern-2
+
+[WSO2 Identity Server Deployment Pattern 2](https://docs.wso2.com/display/IS540/Deployment+Patterns#DeploymentPatterns-Pattern2-HAclustereddeploymentofWSO2IdentityServerwithWSO2IdentityAnalytics) 
+runs two Identity Server instances fronted with a load balancer with WSO2 
+Identity Analytics Server. Identity 
+Server nodes are clustered and a shared file system or a synchronization mechanism like rsync can be used to synchronize
```
[the remark this answers is not in the document]
Comment (malithie) [reply wso2is/hieradata/dev/wso2/wso2is/pattern-2/README.md]: README is formatted looks like how the file is viewed here. It breaks content as it is viewed as a file.

File: wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md (hunk 5 above)
Comment (DilanUA) [reply wso2is_analytics/hieradata/dev/wso2/wso2is_analytics/pattern-1/README.md]: Still, local member ip can be obtained from ipaddress, so, better to keep it here as it is in the hiera yaml.

**Review by DilanUA — CHANGES_REQUESTED** [review]

**Review by DilanUA — APPROVED** [review]

**Pull request merged**